modules/charon-core/src/test/java/org/wso2/charon3/core/protocol/endpoints/UserResourceManagerTest.java

```diff
@@ -115,89 +364,865 @@ public void testGetSuccess(String id, String attributes,
+        abstractResourceManager.when(()
+                -> AbstractResourceManager.encodeSCIMException(any(InternalErrorException.class)))
+                .thenReturn(getEncodeSCIMExceptionObject(new InternalErrorException()));
+        Mockito.when(userManager.getUser(existingId,
+                ResourceManagerUtil.getAllAttributeURIs(schema))).thenReturn(null);
+        SCIMResponse scimResponse = userResourceManager.updateWithPATCH(existingId, scimObjectString,
+                null, attributes, excludeAttributes);
+        Assert.assertEquals(scimResponse.getResponseStatus(), expectedScimResponseStatus);
+    }
+
+    @DataProvider(name = "dataForUpdateWithPATCHReplace")
+    public Object[][] dataToUpdateWithPATCHReplace()
+            throws BadRequestException, CharonException, InternalErrorException {
+
+        SCIMResourceTypeSchema schema = SCIMResourceSchemaManager.getInstance().getUserResourceSchema();
+        JSONDecoder decoder = new JSONDecoder();
+        User userOld = decoder.decodeResource(NEW_USER_SCIM_OBJECT_STRING_FOR_PATCH_WITH_REPLACE, schema, new User());
+        String id = userOld.getId();
+        User userNew = decoder.decodeResource(NEW_USER_SCIM_OBJECT_STRING_FOR_PATCH_WITH_REPLACE_AND_UPDATED,
+                schema, new User());
+        return new Object[][]{
+                {id, NEW_USER_SCIM_OBJECT_STRING_FOR_PATCH_WITH_REPLACE_AND_UPDATED, "userName",
+                        null, userNew, userOld, ResponseCodeConstants.CODE_OK}
+        };
+    }
+
+    @Test(dataProvider = "dataForUpdateWithPATCHReplace")
+    public void testUpdateWithPATCHReplace(String existingId, String scimObjectString,
+                                           String attributes, String excludeAttributes, Object objectNEWUser,
+                                           Object objectOLDUser, int expectedScimResponseStatus)
+            throws BadRequestException, CharonException, NotImplementedException, NotFoundException {
+
+        User userNew = (User) objectNEWUser;
+        User userOld = (User) objectOLDUser;
+        SCIMResourceTypeSchema schema = SCIMResourceSchemaManager.getInstance().getUserResourceSchema();
+        abstractResourceManager.when(() -> AbstractResourceManager.getResourceEndpointURL(SCIMConstants.USER_ENDPOINT))
+                .thenReturn(SCIM2_USER_ENDPOINT);
+        Mockito.when(userManager.getUser(existingId,
+                ResourceManagerUtil.getAllAttributeURIs(schema))).thenReturn(userOld);
+        User validatedUser = (User) ServerSideValidator.validateUpdatedSCIMObject(userOld, userNew, schema);
+        Mockito.when(userManager.updateUser(any(User.class), anyMap(), anyList())).thenReturn(validatedUser);
+        SCIMResponse scimResponse = userResourceManager.updateWithPATCH(existingId, scimObjectString,
+                userManager, attributes, excludeAttributes);
+        Assert.assertEquals(scimResponse.getResponseStatus(), expectedScimResponseStatus);
+    }
```

Comment:
is there a test case for PATCH with remove operation?

Reply:
Fixed in commit 504369f8bdb4f024c63ae5d78f16b04decbedc11